### PR TITLE
Draft of parallel mode Port/socker component

### PR DESCRIPTION
**Pull request by SmallRoomLabs**
( I don't really expect you to actually merge this request even when I'm done with it. )

This is my first draft of the port/socket communication module. <del> It currently lacks any text descriptions so there's a lot of logs about that. And of course the properties popup looks a bit funny.  But those are easy fixes.</del>

Could you please have a quick look at the code and see if I've interpreted how code should be written and how stuff should be implemented in Digital somewhat correctly. (Frankly I'm a bit amazed that I've been able to do this in an unfamiliar language in unfamiliar fairly complex codebase at all...)

I also have some questions about what some of the methods actually do, and also some questions of how to best implement some stuff and it seems like here is a better place to do that than in the issues.   I'll add the questions to this thread a bit later - in the hope that you have some spare time explaining.

### CODE REVIEW

**Comment by SmallRoomLabs** [comment]
Ok, got a question.

I can't get the descriptions from lang_en_xml for the inputs of the component to be picked up. The component itself and the outputs are getting the descriptions just fine.  I can't really figure out what I'm missing here.

```
<string name="elem_Port_pin_C">[Serial mode] This pin expects ....</string>
<string name="elem_Port_pin_ack">[Parallel mode] A rising edge acknowledges ....</string>
<string name="elem_Port_pin_stb">[Parallel mode] A rising edge latches ....</string>
<string name="elem_Port_pin_D">[Parallel mode] The 8 bits of data to be sent....</string>
```

Screen Shot 2019-10-11 at 6 57 19 PM

**Comment by hneemann** [comment]
> But is there a reason that the outputs doesn't need to be told what language to use and the inputs do?

Both, input and output nedds to know which lang keys to use. At the outputs this is done by the `.setPinDescription(DESCRIPTION)` call.

The differences are due to the fact that inputs and outputs in digital are fundamentally different.
An output signal (ObervableValue) is generated directly in the constructor of the node that generates the signal.
With an input it is only defined that a signal is required. Independent of the current instance.
In fact, only output signals are created. These signal classes are then distributed to all nodes as inputs by calling the setInputs() method.

> Could you please have a quick look at the code and see if I've interpreted how code should be written and how stuff should be implemented in Digital somewhat correctly.

There are only two major changes that need to be made:

1. A node in the `readInputs()` method must never write to an output. And never read an input in the `writeOutputs()` method. If a value is to be passed from an input to an output, the value must be stored in a variable in the `readInputs()` method and this stored value needs to be set to the output in the `writeOutputs()` method. Unfortunately, this rule cannot be enforced. The programmer must take care of this himself. If this rule is violated, many hard-to-find errors can occur. This rule guarantees that e.g. two AND gates whose input changes at the same time actually react to it at the same time and not one after the other.

2. When a thread is started in Java, the VM is not terminated until the thread is terminated. If a thread is to run until the VM is terminated, the method `setDaemon(true);` must be called before starting the thread, preferably in the constructor.

File: src/main/java/de/neemann/digital/core/io/Port.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+                        if (--bsyCnt==0) bsy.setValue(0);
+                    }
+                }
+            }
+
+            // If there's pending data in the buffer and the last
+            // data has been ack'ed then update the Q port
+            if (!buf.isEmpty() && !avail.getBool()) {
+                poutVal=buf.remove();
+                pout.set((long) poutVal, 0);
```
Comment (hneemann) [inline src/main/java/de/neemann/digital/core/io/Port.java]: Never write to an output in the readInputs() method.

File: src/main/java/de/neemann/digital/core/io/Port.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+                avail.setValue(1);
+            }
+
+            // If rising edge on ack then clear the avail flag
+            if (lastAck != nowAck) {
+                if (nowAck) avail.setValue(0);
```
Comment (hneemann) [inline src/main/java/de/neemann/digital/core/io/Port.java]: Never write to an output in the readInputs() method.

File: src/main/java/de/neemann/digital/core/io/PortSocket.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * @param port instance of creator
+     * @param listenPort - TCP port to liste to
+     * @param telnetMode - True if a telnet terminal will connect
+     */
+    public PortSocket(Port port, int listenPort, boolean telnetMode) {
+        this.port = port;
+        this.telnetMode = telnetMode;
+        try {
```
Comment (hneemann) [inline src/main/java/de/neemann/digital/core/io/PortSocket.java]: Add the `setDaemon(true);` call.

Comment (SmallRoomLabs) [reply src/main/java/de/neemann/digital/core/io/PortSocket.java]: Ok, added. And after reading the dox it seems like threads started by a daemon thread inherits the status so I don't need to daemonize the actual comms thread after that.

File: src/main/java/de/neemann/digital/core/io/Port.java (hunk 1 above)
Comment (SmallRoomLabs) [reply src/main/java/de/neemann/digital/core/io/Port.java]: That was esoteric knowledge ;-) But seems very logical.

**Comment by SmallRoomLabs** [comment]
Thank you for the comments.  Now I feel (a bit more) confident to implement the serial part as well - as it seems I'm not totally completely off.

**Comment by SmallRoomLabs** [comment]
So... another question.  Is there an easy way to have strings in the setComboBoxValues()? The way it's done in MIDI for the instrument names seems a bit complex for a few static strings.

The setting for Parallel/Serial mode would be better with a combobox compared to the current checkbox.

**Comment by hneemann** [comment]
Use a `Key.KeyEnum` instance. See `Keys.BARREL_SHIFTER_MODE` as an example.
Please pull at first the latest changes to master to avoid "no editor found" errors.

**Comment by SmallRoomLabs** [comment]
Thanks, I'll try the KeyEnum - I see that you just worked on it ;-)  

I hope that my upstream pull, merge and feature branch rebase dir work correctly, it's always scary to do that.  I wish there was a single git command that would to it all automatically as long as there are no conflicts.

**Comment by hneemann** [comment]
> Thanks, I'll try the KeyEnum - I see that you just worked on it ;-)

There was a problem that had annoyed me for some time, but which I had forgotten. Your question reminded me, so I fixed it. :smile:

**Comment by SmallRoomLabs** [comment]
The enums worked fine, but it was a bit annoying to be forced to put the enum in a separate file. I guess that is something one gets used to when doing java devs...

Some more questions:

1) It seems like a component is optimized away from the calculations/propagations unless there has been a change in one of its inputs.  How do I signal that my output has changed internally? Like if I have a component that outputs a random number, or the current temperature of the CPU for instance?

2) <del>Is there an (easy) way to locate an active clock component?</del>
Hehe... While writing this questions I remembered that the monoflop gets evil if there are more than one clock source.  And there I had the answer to my question. So easy.
`ArrayList<Clock> clockList = model.getClocks();`

While thinking of it - it's a bit annoying that you just can have one clock source with the monoflop. Is there any underlying reason that prevents to just use the clock with the highest frequency if there are several?

3) How can I retrieve the name of the currently loaded .dig file?

**Comment by hneemann** [comment]
> It seems like a component is optimized away from the calculations/propagations unless there has been a change in one of its inputs. 

Of course! If it were different, the current performance would not be achievable.

> How do I signal that my output has changed internally?

That's a bit tricky because you have to avoid race conditions. If you want to put e.g. a received character from the receiver thread:

```
public class Test extends Node implements Element {
    private Model model;

    /**
     * Save to be called from any thread.
     *
     * @param c the character received
     */
    public void addCharacterReceived(char c) {
        model.access(() -> putCharToReadQueue(c));
    }

    /**
     * Not save to be called from an arbitrary thread, because parts of the 
     * model are modified.
     *
     * @param c the character received
     */
    private void putCharToReadQueue(char c) {
        // put char to queue or process it otherwise
        
        // after that, notify the model that this node has changed
        hasChanged();
    }

    /**
     * Store the model to synchronize actions
     *
     * @param model the model this element belongs to
     * @throws NodeException NodeException
     */
    @Override
    public void init(Model model) throws NodeException {
        this.model = model;
    }
}
```

> While thinking of it - it's a pity that you just can have one clock source with the monoflop. Is there any underlying reason that prevents to just use the clock with the highest frequency if there are several?

No! There is no reason. Maybe I should implement your suggestion.

> How can I retrieve the name of the currently loaded .dig file?

So far it is not possible to obtain a file name from the model.  What is your use case?

**Comment by hneemann** [comment]
> The enums worked fine, but it was a bit annoying to be forced to put the enum in a separate file. I guess that is something one gets used to when doing java devs...

It is possible to declare an enum inside of an other class. IMO that's the most common way to define an enum. See [here](https://github.com/hneemann/Digital/blob/master/src/main/java/de/neemann/digital/core/element/PinDescription.java).

**Comment by SmallRoomLabs** [comment]
I'll try out that example to get the parallel mode socket work without the clock input. That is also the reason for me (almost) asking about retrieving the clock sources.   In parallel mode I have a 100 cycle "working time" for when transmitting.  I could just as well "auto-use" a clock source, if there's any available in the design.  If there's no clock then the woking-time will be zero. Less pins for the user to connect.

The thing about the monoflops is merely an annoyance, but if there's an easy fix then it might be a good idea. It could save some poor user some aggravation some time.

The user case for retrieving the file name would just be to present it as `[Digital - mydesign.dig]` instead of just `[Digital]` when connecting to it with telnet.   But when thinking about it this banner is only displayed at the  initial connect - the session is never dropped until Digital is terminated.

OK, I'll try again to bake in the enum into the Port.java - as I remember it the linter complained about that it should be in its own file when I tried the first time and I just followed the hint. :-)

**Comment by SmallRoomLabs** [comment]
There's no need to keep this open, so I'll close it now.

**Pull request closed without merging**